File: surface/widgets/swiftui/components/widgets/concepts/process-foundation/VariableInspector.swift
import SwiftUI

enum VariableInspectorWidgetState {
    case idle, filtering, varSelected
}

enum VariableInspectorEvent {
    case search, selectVar, addWatch, clear, deselect
}

func variableInspectorReduce(state: VariableInspectorWidgetState, event: VariableInspectorEvent) -> VariableInspectorWidgetState {
    switch state {
    case .idle:
        if event == .search { return .filtering }
        if event == .selectVar { return .varSelected }
        return state
    case .filtering:
        if event == .clear { return .idle }
        if event == .selectVar { return .varSelected }
        return state
    case .varSelected:
        if event == .deselect { return .idle }
        if event == .selectVar { return .varSelected }
        return state
    }
}

struct ProcessVariable: Identifiable {
    var id: String { name }
    var name: String
    var type: String
    var value: String
    var scope: String? = nil
    var changed: Bool = false
}

struct WatchExpression: Identifiable {
    var id: String
    var expression: String
    var value: String? = nil
}

struct VariableInspectorView: View {
    var variables: [ProcessVariable]
    var runStatus: String
    var showTypes: Bool = true
    var showWatch: Bool = true
    var expandDepth: Int = 1
    var watchExpressions: [WatchExpression] = []
    var onSelectVariable: ((String) -> Void)? = nil
    var onAddWatch: ((String) -> Void)? = nil
    var onRemoveWatch: ((String) -> Void)? = nil

    @State private var widgetState: VariableInspectorWidgetState = .idle
    @State private var searchQuery: String = ""
    @State private var selectedVar: String? = nil
    @State private var focusIndex: Int = 0

    private func typeBadge(_ type: String) -> String {
        let map = ["string": "str", "number": "num", "boolean": "bool", "object": "obj", "array": "arr"]
        return map[type.lowercased()] ?? type
    }

    private var filteredVariables: [ProcessVariable] {
        if searchQuery.isEmpty { return variables }
        let q = searchQuery.lowercased()
        return variables.filter { $0.name.lowercased().contains(q) }
    }

    var body: some View {
        VStack(alignment: .leading, spacing: 8) {
            // Search bar
            HStack {
                TextField("Filter variables...", text: $searchQuery)
                    .textFieldStyle(.roundedBorder)
                    .accessibilityLabel("Filter variables by name")
                    .onChange(of: searchQuery) { newVal in
                        if newVal.isEmpty {
                            widgetState = variableInspectorReduce(state: widgetState, event: .clear)
                        } else {
                            widgetState = .idle
                            widgetState = variableInspectorReduce(state: widgetState, event: .search)
                        }
                    }

                if !searchQuery.isEmpty {
                    Button {
                        searchQuery = ""
                        widgetState = variableInspectorReduce(state: widgetState, event: .clear)
                    } label: {
                        Image(systemName: "xmark.circle.fill")
                            .foregroundColor(.secondary)
                    }
                    .accessibilityLabel("Clear search")
                }
            }

            // Variable list
            if filteredVariables.isEmpty {
                Text(searchQuery.isEmpty ? "No variables available" : "No variables match the filter")
                    .font(.body)
                    .foregroundColor(.secondary)
                    .frame(maxWidth: .infinity)
                    .padding()
            } else {
                LazyVStack(spacing: 4) {
                    ForEach(filteredVariables) { variable in
                        HStack(spacing: 8) {
                            Text(variable.name)
                                .font(.system(.body, design: .monospaced))
                                .fontWeight(selectedVar == variable.name ? .bold : .regular)

                            if showTypes {
                                Text(typeBadge(variable.type))
                                    .font(.caption2)
                                    .padding(.horizontal, 6)
                                    .padding(.vertical, 2)
                                    .background(Color.secondary.opacity(0.15))
                                    .cornerRadius(4)
                            }

                            if let scope = variable.scope {
                                Text(scope)
                                    .font(.caption2)
                                    .foregroundColor(.secondary)
                                    .accessibilityLabel("Scope: \(scope)")
                            }

                            Spacer()

                            Text(variable.value)
                                .font(.system(.caption, design: .monospaced))
                                .lineLimit(1)
                                .foregroundColor(.secondary)

                            if variable.changed {
                                Circle()
                                    .fill(Color.blue)
                                    .frame(width: 6, height: 6)
                                    .accessibilityLabel("Value changed")
                            }
                        }
                        .padding(.horizontal, 8)
                        .padding(.vertical, 6)
                        .background(
                            RoundedRectangle(cornerRadius: 4)
                                .fill(selectedVar == variable.name ? Color.accentColor.opacity(0.1) : Color.clear)
                        )
                        .contentShape(Rectangle())
                        .onTapGesture {
                            selectedVar = variable.name
                            widgetState = variableInspectorReduce(state: .idle, event: .selectVar)
                            onSelectVariable?(variable.name)
                        }
                        .accessibilityLabel("\(variable.name): \(variable.value)")
                    }
                }
            }

            // Watch list
            if showWatch {
                VStack(alignment: .leading, spacing: 6) {
                    HStack {
                        Text("Watch Expressions")
                            .font(.subheadline)
                            .fontWeight(.semibold)
                        Spacer()
                        Button("+ Watch") {
                            if let name = selectedVar, !name.isEmpty {
                                onAddWatch?(name)
                            }
                        }
                        .font(.caption)
                        .accessibilityLabel("Add watch expression")
                    }

                    ForEach(watchExpressions) { watch in
                        HStack {
                            Text(watch.expression)
                                .font(.system(.caption, design: .monospaced))
                            Spacer()
                            Text(watch.value ?? "evaluating...")
                                .font(.system(.caption, design: .monospaced))
                                .foregroundColor(.secondary)
                            Button {
                                onRemoveWatch?(watch.id)
                            } label: {
                                Image(systemName: "xmark")
                                    .font(.caption2)
                                    .foregroundColor(.secondary)
                            }
                            .accessibilityLabel("Remove watch: \(watch.expression)")
                        }
                        .padding(.horizontal, 8)
                        .padding(.vertical, 4)
                    }
                }
                .padding(8)
                .overlay(
                    RoundedRectangle(cornerRadius: 6)
                        .stroke(Color.gray.opacity(0.2), lineWidth: 1)
                )
            }
        }
        .accessibilityElement(children: .contain)
        .accessibilityLabel("Variable inspector")
    }
}

File: surface/widgets/compose/components/widgets/concepts/process-foundation/VariableInspector.kt
package com.clef.surface.widgets.concepts.processfoundation

import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.itemsIndexed
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp

// --- State machine ---

enum class VariableInspectorState { Idle, Filtering, VarSelected }

sealed class VariableInspectorEvent {
    data class Search(val query: String) : VariableInspectorEvent()
    data class SelectVar(val name: String) : VariableInspectorEvent()
    data class AddWatch(val name: String) : VariableInspectorEvent()
    object Clear : VariableInspectorEvent()
    object Deselect : VariableInspectorEvent()
}

fun variableInspectorReduce(state: VariableInspectorState, event: VariableInspectorEvent): VariableInspectorState = when (state) {
    VariableInspectorState.Idle -> when (event) {
        is VariableInspectorEvent.Search -> VariableInspectorState.Filtering
        is VariableInspectorEvent.SelectVar -> VariableInspectorState.VarSelected
        is VariableInspectorEvent.AddWatch -> VariableInspectorState.Idle
        else -> state
    }
    VariableInspectorState.Filtering -> when (event) {
        is VariableInspectorEvent.Clear -> VariableInspectorState.Idle
        is VariableInspectorEvent.SelectVar -> VariableInspectorState.VarSelected
        else -> state
    }
    VariableInspectorState.VarSelected -> when (event) {
        is VariableInspectorEvent.Deselect -> VariableInspectorState.Idle
        is VariableInspectorEvent.SelectVar -> VariableInspectorState.VarSelected
        else -> state
    }
}

// --- Types ---

data class ProcessVariable(
    val name: String,
    val type: String,
    val value: String,
    val scope: String? = null,
    val changed: Boolean = false
)

data class WatchExpression(
    val id: String,
    val expression: String,
    val value: String? = null
)

// --- Helpers ---

private fun typeBadge(type: String): String = when (type.lowercase()) {
    "string" -> "str"
    "number" -> "num"
    "boolean" -> "bool"
    "object" -> "obj"
    "array" -> "arr"
    else -> type
}

@Composable
fun VariableInspector(
    variables: List<ProcessVariable>,
    runStatus: String,
    modifier: Modifier = Modifier,
    showTypes: Boolean = true,
    showWatch: Boolean = true,
    watchExpressions: List<WatchExpression> = emptyList(),
    onSelectVariable: (String) -> Unit = {},
    onAddWatch: (String) -> Unit = {},
    onRemoveWatch: (String) -> Unit = {}
) {
    var state by remember { mutableStateOf(VariableInspectorState.Idle) }
    var searchQuery by remember { mutableStateOf("") }
    var selectedVar by remember { mutableStateOf<String?>(null) }

    val filteredVariables = remember(variables, searchQuery) {
        if (searchQuery.isBlank()) variables
        else variables.filter { it.name.contains(searchQuery, ignoreCase = true) }
    }

    Column(modifier = modifier.semantics { contentDescription = "Variable inspector" }) {
        // Search bar
        OutlinedTextField(
            value = searchQuery,
            onValueChange = {
                searchQuery = it
                state = if (it.isNotBlank()) variableInspectorReduce(state, VariableInspectorEvent.Search(it))
                else variableInspectorReduce(state, VariableInspectorEvent.Clear)
            },
            placeholder = { Text("Filter variables\u2026") },
            singleLine = true,
            modifier = Modifier.fillMaxWidth().padding(bottom = 8.dp),
            trailingIcon = {
                if (searchQuery.isNotEmpty()) {
                    IconButton(onClick = {
                        searchQuery = ""
                        state = variableInspectorReduce(state, VariableInspectorEvent.Clear)
                    }) { Text("\u2715", fontSize = 14.sp) }
                }
            }
        )

        // Variable list
        LazyColumn(modifier = Modifier.weight(1f)) {
            itemsIndexed(filteredVariables) { _, variable ->
                val isSelected = selectedVar == variable.name
                Row(
                    verticalAlignment = Alignment.CenterVertically,
                    modifier = Modifier
                        .fillMaxWidth()
                        .clickable {
                            selectedVar = variable.name
                            state = variableInspectorReduce(state, VariableInspectorEvent.SelectVar(variable.name))
                            onSelectVariable(variable.name)
                        }
                        .padding(vertical = 6.dp, horizontal = 4.dp)
                ) {
                    // Name
                    Text(
                        variable.name,
                        fontWeight = if (isSelected) FontWeight.Bold else FontWeight.Normal,
                        fontSize = 13.sp,
                        fontFamily = FontFamily.Monospace,
                        modifier = Modifier.weight(0.3f)
                    )

                    // Type badge
                    if (showTypes) {
                        Surface(tonalElevation = 2.dp, shape = MaterialTheme.shapes.small) {
                            Text(
                                typeBadge(variable.type),
                                fontSize = 10.sp,
                                modifier = Modifier.padding(horizontal = 4.dp, vertical = 1.dp)
                            )
                        }
                        Spacer(Modifier.width(6.dp))
                    }

                    // Scope
                    if (variable.scope != null) {
                        Text(variable.scope, fontSize = 10.sp, color = MaterialTheme.colorScheme.onSurfaceVariant)
                        Spacer(Modifier.width(6.dp))
                    }

                    // Value
                    Text(
                        variable.value,
                        fontSize = 12.sp,
                        fontFamily = FontFamily.Monospace,
                        color = MaterialTheme.colorScheme.onSurfaceVariant,
                        modifier = Modifier.weight(0.5f),
                        maxLines = 1
                    )

                    // Changed indicator
                    if (variable.changed) {
                        Text("\u2022", fontSize = 14.sp, color = MaterialTheme.colorScheme.primary)
                    }
                }
                HorizontalDivider()
            }

            if (filteredVariables.isEmpty()) {
                item {
                    Text(
                        if (searchQuery.isNotEmpty()) "No variables match the filter" else "No variables available",
                        modifier = Modifier.padding(12.dp),
                        color = MaterialTheme.colorScheme.onSurfaceVariant
                    )
                }
            }
        }

        // Watch list
        if (showWatch) {
            HorizontalDivider(modifier = Modifier.padding(vertical = 8.dp))
            Row(
                verticalAlignment = Alignment.CenterVertically,
                horizontalArrangement = Arrangement.SpaceBetween,
                modifier = Modifier.fillMaxWidth()
            ) {
                Text("Watch Expressions", fontWeight = FontWeight.SemiBold, fontSize = 13.sp)
                TextButton(onClick = {
                    val expr = selectedVar ?: ""
                    if (expr.isNotEmpty()) {
                        state = variableInspectorReduce(state, VariableInspectorEvent.AddWatch(expr))
                        onAddWatch(expr)
                    }
                }) {
                    Text("+ Watch", fontSize = 12.sp)
                }
            }

            watchExpressions.forEach { watch ->
                Row(
                    verticalAlignment = Alignment.CenterVertically,
                    modifier = Modifier.fillMaxWidth().padding(vertical = 2.dp)
                ) {
                    Text(watch.expression, fontFamily = FontFamily.Monospace, fontSize = 12.sp, modifier = Modifier.weight(1f))
                    Text(
                        watch.value ?: "evaluating\u2026",
                        fontSize = 12.sp,
                        color = MaterialTheme.colorScheme.onSurfaceVariant,
                        modifier = Modifier.weight(1f)
                    )
                    IconButton(onClick = { onRemoveWatch(watch.id) }, modifier = Modifier.size(24.dp)) {
                        Text("\u2715", fontSize = 12.sp)
                    }
                }
            }
        }
    }
}
